### ISSUE TITLE
realm->_realm->thread_id() != std::this_thread::get_id() when RLMGetThreadLocalCachedRealmForPath()

### ISSUE THREAD

**Issue by belakva** [realm/realm-cocoa#4132]
We have a heavily async app. Occasionally after say 20 minutes of active user interaction we get `IncorrectThreadException()` in different methods though we never pass Realm instances between threads.

It appeared that these crashes occur when `RLMGetThreadLocalCachedRealmForPath()` returns Realm with `thread_id()`, which is not equal to `std::this_thread::get_id()`.

Therefore we suggest additional check that thread ids match, probably before `RLMCacheRealm()`, or before returning Realm got with `RLMGetThreadLocalCachedRealmForPath()` or both.

[Related Issue](https://github.com/realm/realm-cocoa/issues/4127)

**Comment by jpsim** [comment]
@belakva can you share a sample project that triggers this issue?

**Comment by belakva** [comment]
@jpsim unfortunately, no. It is hard to reproduce the issue - it might be, crashes like that happened before in our app from time to time, but only after heavy increase of interaction with Realm, we faced regular crashes. Our app uses Objective-C and ReactiveCocoa, and crashes appear usually in `RACScheduler` blocks. So I do not see the possibility of ripping off a part of an app to use as a sample project, sorry.

It is important to understand why Realm instance with wrong `thread_id()` is written to `NSMapTable`.

**Comment by jpsim** [comment]
I appreciate that it might be difficult to extract a reproducible sample. I'm parking this until someone (either on our team, you, or someone else) can manage to do so.

**Comment by austinzheng** [comment]
I believe this was fixed in 2.2.0 (`Fix spurious incorrect thread exceptions when a thread id happens to be reused while an RLMRealm instance from the old thread still exists.`). If this issue persists after upgrading to the latest version of Realm, please file a new ticket with appropriate reproduction information.
